[IOPLT-748] Add listTrials to OpenAPI Specification.

robbcocco opened the pull request:
<!---
Please always add a PR description as if nobody knows anything about the context
these changes come from. Even if we are all from our internal team, we may not
be on the same page. Write this PR as you were contributing to a public OSS
project, where nobody knows you and you have to earn their trust. This will
improve our projects in the long run!

Thanks :).
-->

#### List of Changes
<!--- Describe your changes in detail -->
- Add `listTrials` to path `GET /trials`
- Add query parameters `queryPageSize`, `queryMaximumId` and `queryMinimumId`
- Add `TrialPaginatedCollection`, `TrialCollection`, `TrialSlim`, `ItemId`, `PreviousId`, `NextId` and `PaginatedCollection` entities

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
To expose the listTrials function and to define the parameters and the return type of said function.

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

#### Checklist before requesting a review
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have performed a self-review of my code.
- [x] I have committed only changes relevant to the task.
- [x] I have minimized the number of changes.
- [x] My changes are about only one task or issue.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

robbcocco committed:
Add listTrials definition to OpenAPI specifications.

robbcocco committed:
Add query parameters for pagination to listTrials.

robbcocco committed:
Add lighter Trial schema for listTrials.
Update Trials schema name to TrialCollection.

kin0992 commented on apps/functions-subscription/api/internal.yaml:
```diff
@@ -196,6 +196,30 @@ paths:
+      description: List all trials
+      summary: List all trials
+      parameters:
+        - $ref: '#/components/parameters/queryPageNumber'
+        - $ref: '#/components/parameters/queryPageSize'
```
**issue(blocking)**: use cursor pagination

robbcocco replied:
Resolved in 728596289119be614fff33a07973a5f8e80d6b63

robbcocco committed:
Swap pagination with cursor navigation.
Update description for listTrial.

robbcocco committed:
Add error 400 to listTrials.

robbcocco commented on apps/functions-subscription/api/internal.yaml:
```diff
@@ -196,6 +196,30 @@ paths:
+    get:
+      operationId: listTrials
+      description: List all trials
+      summary: List all trials
```
Resolved in 728596289119be614fff33a07973a5f8e80d6b63

kin0992 commented on apps/functions-subscription/api/internal.yaml:
```diff
@@ -196,6 +196,37 @@ paths:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TrialCollection'
```
**thought**: how the caller can handle the pagination?

assume we have 100 documents, the `pageSize` query parameter is by default `25`, so the HTTP call will return 25 documents.

do you think would be better to have the next id (the id to use to fetch the next batch of documents) within the response?

robbcocco replied:
Added nextId to response in 71323943e0b5b6de4d732d5465c4bafa2973e4b1

kin0992 commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -474,3 +538,19 @@ components:
                   type: string
                   description: The identity ID used to receive events from the queue.
                   minLength: 1
+    TrialForCollection:
```
**question(non-blocking)**: what do you think of a name like `TrialSlim` which represents a slim version of the `Trial`?

Anyway, it's up to you decide what naming to use :)

robbcocco replied:
Sounds good! Honestly I'm not that good at coming up with new names ahah

kin0992 commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -273,6 +304,35 @@ components:
+        minimum: 1
+        maximum: 100
+        default: 25
+    queryMaximumId:
```
**suggestion**: add a description

I'd a description about what means this parameter and how a user can used.

robbcocco replied:
Added descriptions in 71323943e0b5b6de4d732d5465c4bafa2973e4b1

kin0992 commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -273,6 +304,35 @@ components:
+        type: string
+        maxLength: 26
+        minLength: 26
+        pattern: '[0-9a-hjkmnp-tv-zA-HJKMNP-TV-Z]{26}'
```
**suggestion**: i don't think this is needed

robbcocco replied:
Removed regex validation in 71323943e0b5b6de4d732d5465c4bafa2973e4b1

robbcocco committed:
Add nextId to schema response for listTrials.
Add schema for NextTrialId.
Update schema names.
Add descriptions.

kin0992 reviewed (commented):
left some comments, but it looks good!

robbcocco committed:
Rename schema TrialCollectionWithNextId to TrialCollectionPaginated.
Add required property to schema's items.

robbcocco commented on apps/functions-subscription/api/internal.yaml:
```diff
@@ -455,6 +520,17 @@ components:
       enum:
         - CREATING
         - CREATED
+    TrialCollectionWithNextId:
```
Renamed into `TrialCollectionPaginated` to stay consistent with other schemas in 250a2da93784e905dc7bc98893f223fa4097ad49

robbcocco commented on apps/functions-subscription/api/internal.yaml:
```diff
@@ -455,6 +520,17 @@ components:
+    TrialCollectionWithNextId:
+      type: object
+      properties:
+        data:
```
Added required property in 250a2da93784e905dc7bc98893f223fa4097ad49

robbcocco committed:
Merge branch 'main' into feature/add-listtrials-to-openapi-specification

robbcocco commented on apps/functions-subscription/api/internal.yaml:
```diff
@@ -196,6 +196,37 @@ paths:
+      operationId: listTrials
+      description: List trials
+      summary: List trials
+      parameters:
+        - $ref: '#/components/parameters/queryPageSize'
+        - $ref: '#/components/parameters/queryMaximumId'
+        - $ref: '#/components/parameters/queryMinimumId'
```
Resolved with merge from main in 4e90f655281b3696ad7ab1cfb034ef482f04d201

robbcocco committed:
Create early-bobcats-learn.md

kin0992 commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -439,6 +504,19 @@ components:
       enum:
         - CREATING
         - CREATED
+    TrialCollectionPaginated:
```
**suggestion(non-blocking)**: we could create a schema only for paginated response

since in OpenAPI you can use multiple references to define a schema, maybe we could have a schema only for pagination (with `next` and `prev` values) and where there are paginated responses, we can reference this new schema and the schema of the items. 

what do you think?

robbcocco replied:
Created PaginatedCollection schema in 20f69ea2658432a88daecb105c2db8f5f1408cac

kin0992 commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -315,6 +376,10 @@ components:
       type: string
       description: Unique identifier of the trial.
       minLength: 1
+    NextTrialId:
+      type: string
+      description: Unique identifier of the next trial.
+      minLength: 1
```
**suggestion**: we can call it just `NextId` and reuse the schema for other endpoints in future

robbcocco replied:
Renamed to nextId in 20f69ea2658432a88daecb105c2db8f5f1408cac

robbcocco committed:
Separate fields for pagination from TrialPaginatedCollection schema.
Rename schema NextTrialId into NextId.
Create PaginatedCollection schema.
Add PreviousId schema.

robbcocco commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -439,6 +504,19 @@ components:
+        items:
+          $ref: '#/components/schemas/TrialCollection'
+        nextId:
+          $ref: '#/components/schemas/NextTrialId'
```
Added previousId in 20f69ea2658432a88daecb105c2db8f5f1408cac

AleDore commented on infra/resources/modules/commons/api/ts_management/v1/_openapi.yaml:
```diff
@@ -439,6 +500,36 @@ components:
       enum:
         - CREATING
         - CREATED
+    NextId:
+      type: string
+      description: Unique identifier of the next item.
+      minLength: 1
+    PreviousId:
```
can we define id with a more strict typings? i.e: 
```
maxLength: 26
minLength: 26
```

robbcocco replied:
since we now have 3 schemas (trialId, nextId and previousId) with the same type, would it be ok to create a schema (ItemId) from which the formers will inherit the typing?
There's a comment in row 296 (Using the reference ($ref) to UserId breaks the client generator) but it seems that using $ref in a schema doesn't break `yarn generate`.

AleDore replied:
yes it can be used as a common schema

robbcocco replied:
resolved in 2b5330db8f4707a14c57fc0cd171946bc6a7988d and 35d9506f1e1ad11c925b87f4b10e7501f4236571 (stricter typing in TrialId breaks the build)

robbcocco committed:
Update TrialId, NextId and PreviousId with stricter typings.

robbcocco committed:
Revert TrialId typing.

AleDore reviewed (approved):
lgtm
